src/mame/drivers/naomi.cpp

```diff
@@ -2635,15 +2035,63 @@ static INPUT_PORTS_START( naomi_mp )
 	PORT_BIT( 0x80, IP_ACTIVE_HIGH, IPT_MAHJONG_D )
 INPUT_PORTS_END
 
+CUSTOM_INPUT_MEMBER(naomi_state::suchie3_mp_r)
+{
+	uint8_t retval = 0;
+
+	int port = 0;
+	for (int i = 0x80; i >= 0x08; i >>= 1, port++)
+	{
+		if (m_mp_mux & i)
+		{
+			// KEY1 and KEY5 are swapped
+			// FIXME: should really be mapped in input struct instead
+			if (port == 0)
+				retval |= m_mp[4].read_safe(0);
+			else if (port == 4)
+				retval |= m_mp[0].read_safe(0);
+			else
+				retval |= m_mp[port].read_safe(0);
+		}
+	}
+	return retval;
+}
+
 static INPUT_PORTS_START( suchie3 )
 	PORT_INCLUDE( naomi_mp )
 	PORT_MODIFY("P1")
 	PORT_BIT( 0xff00, IP_ACTIVE_HIGH, IPT_CUSTOM ) PORT_CUSTOM_MEMBER(naomi_state, suchie3_mp_r)
 INPUT_PORTS_END
 
+template <int P> CUSTOM_INPUT_MEMBER(naomi_state::naomi_kb_r)
+{
+	u8 retval = 0;
+
+	for(int i = 0; i < 5; i++)
+	{
+		uint32_t row;
+
+		// read the current row
+		row = ((P == 0) ? m_p1_kb : m_p2_kb)[i]->read();
+
+		// if anything is pressed, convert the 32-bit raw value to keycode
+		if(row != 0)
+		{
+			// base value x20
+			retval = i * 0x20;
+			for(int j = 0; j < 32; j++)
+			{
+				if(row & 1 << j)
+					return retval + j;
+			}
```

Comment:
You’re doing `for(...)` and `if(...)` here when the rest of the file does `for (...)` and `if (...)`

Also, the use of `retval` is weird – it can only be assigned a non-zero value inside here, and there’s no way to fall off the end of the loop if it’s assigned a non-zero value.  It would be a lot cleared to declare `retval` inside the scope where it’s actually used and `return 0` at the end of the function.  Having it the way it is suggests `retval` may carry a meaningful value across loop iterations, but it doesn’t.

Reply:
Not really touched this part but yes, I see your point here.

Reply:
Oh was that moved from machine/naomi.cpp?  It’s a bit confusing working out what’s just shuffled and what’s changed.